concordium-cis2/src/cis2_client.rs

```diff
@@ -0,0 +1,724 @@
+//! CIS2 contract for the following methods
+//! - `supports_cis2` : Calls [`supports`](https://proposals.concordium.software/CIS/cis-0.html#supports)
+//! - `operator_of` : Calls [`operatorOf`](https://proposals.concordium.software/CIS/cis-2.html#operatorof)
+//! - `balance_of` : Calls [`balanceOf`](https://proposals.concordium.software/CIS/cis-2.html#balanceof)
+//! - `transfer` : Calls [`transfer`](https://proposals.concordium.software/CIS/cis-2.html#transfer)
+//! - `update_operator` : Calls [`updateOperator`](https://proposals.concordium.software/CIS/cis-2.html#updateoperator)
+
+use crate::*;
+use concordium_std::*;
+
+const SUPPORTS_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("supports");
+const OPERATOR_OF_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("operatorOf");
+const BALANCE_OF_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("balanceOf");
+const TRANSFER_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("transfer");
+const UPDATE_OPERATOR_ENTRYPOINT_NAME: EntrypointName =
+    EntrypointName::new_unchecked("updateOperator");
+
+/// Wrapper for `Cis2Error<T>`. This allows for various traits to be
+/// implemented.
+#[derive(Debug)]
+pub struct Cis2ErrorWrapper<T>(Cis2Error<T>);
+
+impl<T> From<Cis2Error<T>> for Cis2ErrorWrapper<T> {
+    fn from(e: Cis2Error<T>) -> Self { Cis2ErrorWrapper(e) }
+}
+
+impl<T> AsRef<Cis2Error<T>> for Cis2ErrorWrapper<T> {
+    fn as_ref(&self) -> &Cis2Error<T> { &self.0 }
+}
+
+impl<T: Serial> Serial for Cis2ErrorWrapper<T> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> { self.0.serial(out) }
+}
+
+impl<T: Deserial> Deserial for Cis2ErrorWrapper<T> {
+    fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
+        let tag = source.read_u8()?;
+        match tag {
+            0 => Ok(Cis2ErrorWrapper(Cis2Error::InvalidTokenId)),
+            1 => Ok(Cis2ErrorWrapper(Cis2Error::InsufficientFunds)),
+            2 => Ok(Cis2ErrorWrapper(Cis2Error::Unauthorized)),
+            3 => {
+                let t = T::deserial(source)?;
+                Ok(Cis2ErrorWrapper(Cis2Error::Custom(t)))
+            }
+            _ => bail!(ParseError {}),
+        }
+    }
+}
+pub type InvokeContractError<T> = CallContractError<Cis2ErrorWrapper<T>>;
+
+/// Errors which can be returned by the `Cis2Client`.
+#[derive(Debug)]
+pub enum Cis2ClientError<T> {
+    /// When there is an error invoking the CIS2 contract.
+    InvokeContractError(InvokeContractError<T>),
+    /// When there is an error parsing the result.
+    ParseResult,
+    /// When the response is invalid. Ex. When the response is empty vector for
+    /// a single query.
+    InvalidResponse,
+}
+
+impl<T: Serial> Serial for Cis2ClientError<T> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        match self {
+            Cis2ClientError::InvokeContractError(e) => {
+                out.write_u8(3)?;
```

Comment:
We are not using the tag `2` for serializing any of the other variants, so I would prefer to use this here
```suggestion
                out.write_u8(2)?;
```

Reply:
done